### PR TITLE
Dense vector field types are indexed by default

### PR DESCRIPTION
**Pull request by carlosdelest**
`dense_vector` field types will be indexed by default, and `similarity` set to `cosine`.

### CODE REVIEW

**Comment by benwtrent** [comment]
I think you accidentally added a bunch of synonym things :). Please remove all those changes.

**Comment by carlosdelest** [comment]
> I think you accidentally added a bunch of synonym things :)

Thanks @benwtrent - this started as something I wanted to check while I was doing something else, and I forgot where it all started 😅 

It should be fixed now!

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 1]
```diff
@@ -128,18 +122,41 @@ public static class Builder extends FieldMapper.Builder {
...
-            this.indexed.requiresParameter(similarity);
-            this.similarity.setSerializerCheck((id, ic, v) -> v != null);
-            this.similarity.requiresParameter(indexed);
-            this.indexOptions.requiresParameter(indexed);
+            boolean indexedByDefault = indexVersionCreated.onOrAfter(INDEXED_BY_DEFAULT_INDEX_VERSION);
+            this.similarity = Parameter.enumParam(
+                "similarity",
+                false,
+                m -> toType(m).similarity,
+                indexedByDefault ? VectorSimilarity.COSINE : null,
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:138]: what happens if somebody passes `index: false` is the vector similarity still set in that case? Seems like we shouldn't do that.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:138]: It should only be `cosine` if `index: true`

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 2]
```diff
@@ -128,18 +122,41 @@ public static class Builder extends FieldMapper.Builder {
...
+            this.indexed = Parameter.indexParam(m -> toType(m).indexed, indexedByDefault).addValidator(v -> {
+                if (v == false) {
+                    if (similarity.isConfigured()) {
+                        throw new IllegalArgumentException(
+                            "Field [similarity] can only be specified for a field of type [dense_vector] when it is indexed"
+                        );
+                    }
+                    if (indexOptions.isConfigured()) {
+                        throw new IllegalArgumentException(
+                            "Field [indexOptions] can only be specified for a field of type [dense_vector] when it is indexed"
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:150]:
```suggestion
                            "Field [index_options] can only be specified for a field of type [dense_vector] when it is indexed"
```

**Review by benwtrent — COMMENTED** [review]
once this moves out of draft, I will review further.

Please add yaml tests as well.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml [hunk 3]
```diff
@@ -5,9 +5,6 @@ setup:
 
 ---
 "Create synonyms with no ID creates ID automatically":
-  - skip:
```
Comment (benwtrent) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml:8]: Could you take all synonym changes and put them in a separate PR? They snuck in again maybe?

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 4]
```diff
@@ -993,14 +993,15 @@ public static <T extends Enum<T>> Parameter<T> restrictedEnumParam(
             String name,
             boolean updateable,
             Function<FieldMapper, T> initializer,
-            T defaultValue,
+            Supplier<T> defaultValue,
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:996]: We need consistency between all our mapper default values. I do not like having a specialized constructor here that takes a supplier but none of the others do.

I need to think a bit more. I would prefer not having supplied default value like this.

Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:996]: I understand. The reasons for adding this were:
- We need to change the default value for similarity based on the value for `index`.
- Other mapper builders provide this as well (indexParam / boolParam, analyzerParam).

Are there better ways to achieve this? I can give it a try 👍

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:996]: > Other mapper builders provide this as well (indexParam / boolParam, analyzerParam).

As long as we are unified with the builders, it could be ok. I will read them. My concern is that somebody passes "null" and instead of it being "() -> null" is `null` and we call `get()` on `null`

Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:996]: Yep, that happened 😢 . I added an `assert` to try and catch that as early as possible, but this could definitely bite us.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:996]: @carlosdelest I would suggest a NEW method/ctor that works with enums that has a supplier. This way if somebody is passing `null`, its a compilation error as Java isn't sure which one you are specifying. That is way better than a runtime error.

Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:1037]: Good call - adding that. Now we need to specify what method are we using to avoid surprises. Thanks for the suggestion! 🙇

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 5]
```diff
@@ -127,20 +122,42 @@ public static class Builder extends FieldMapper.Builder {
...
+            final boolean indexedByDefault = indexVersionCreated.onOrAfter(INDEXED_BY_DEFAULT_INDEX_VERSION);
+            this.indexed = Parameter.indexParam(m -> toType(m).indexed, indexedByDefault).alwaysSerialize();
+            this.similarity = Parameter.enumParam(
+                "similarity",
+                false,
+                m -> toType(m).similarity,
+                (Supplier<VectorSimilarity>) () -> indexedByDefault && indexed.getValue() ? VectorSimilarity.COSINE : null,
+                VectorSimilarity.class
+            ).acceptsNull().setSerializerCheck((id, ic, v) -> v != null);
+            this.indexed.addValidator(v -> {
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:143]: Switched to using validators, as validation now depends on both the values (and defaults) and the configured parameters.

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java [hunk 6]
```diff
@@ -77,8 +77,9 @@ protected void minimalMapping(XContentBuilder b) throws IOException {
         if (elementType != ElementType.FLOAT) {
             b.field("element_type", elementType.toString());
         }
+        b.field("index", indexed);
```
Comment (carlosdelest) [inline server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:80]: This isn't the minimal mapping - we could do without the `index` field. However, this would make `testDeprecatedBoost` fail, as it uses `minimalMapping` to create a mapping where index defaults to false.

We could create an additional method to generate the mapping for `testDeprecatedBoost`, but seemed overkill. The actual minimal params are tested in this class.

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java [hunk 7]
```diff
@@ -435,31 +452,83 @@ public void testInvalidParameters() {
...
+            b.field("type", "dense_vector").field("dims", 3).field("index", true).field("similarity", "dot_product");
+        }));
+        DenseVectorFieldMapper denseVectorFieldMapper = (DenseVectorFieldMapper) documentMapper.mappers().getMapper("field");
+        DenseVectorFieldType denseVectorFieldType = denseVectorFieldMapper.fieldType();
+
+        assertTrue(denseVectorFieldType.isIndexed());
+        assertEquals(VectorSimilarity.DOT_PRODUCT, denseVectorFieldType.getSimilarity());
+    }
+
+    public void testDefaultParamsIndexByDefault() throws Exception {
```
Comment (carlosdelest) [inline server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:517]: Default params tested here, so we don't rely on `minimalMapping()` for that.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 8]
```diff
@@ -974,6 +974,25 @@ public static <T extends Enum<T>> Parameter<T> enumParam(
             Function<FieldMapper, T> initializer,
             T defaultValue,
             Class<T> enumClass
+        ) {
+            return enumParam(name, updateable, initializer, (Supplier<T>) () -> defaultValue, enumClass);
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:978]: Added extra `Supplier<T>` related methods here, so we are aware of what version we're using (see https://github.com/elastic/elasticsearch/pull/98268/files#r1287069780)

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/80_dense_vector_indexed_by_default.yml [hunk 9]
```diff
@@ -0,0 +1,133 @@
+setup:
+  - skip:
+      version: ' - 8.9.99'
+      reason: 'dense_vector indexed by default was added in 8.10'
+
+---
+"Indexed by default with cosine similarity":
```
Comment (carlosdelest) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/80_dense_vector_indexed_by_default.yml:7]: Added YAML tests - some of them are already tested in unit tests, LMKWYT about removing incorrect params tests

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 10]
```diff
@@ -127,20 +122,42 @@ public static class Builder extends FieldMapper.Builder {
...
             super(name);
             this.indexVersionCreated = indexVersionCreated;
-
-            this.indexed.requiresParameter(similarity);
-            this.similarity.setSerializerCheck((id, ic, v) -> v != null);
-            this.similarity.requiresParameter(indexed);
-            this.indexOptions.requiresParameter(indexed);
-            this.indexOptions.setSerializerCheck((id, ic, v) -> v != null);
+            final boolean indexedByDefault = indexVersionCreated.onOrAfter(INDEXED_BY_DEFAULT_INDEX_VERSION);
+            this.indexed = Parameter.indexParam(m -> toType(m).indexed, indexedByDefault).alwaysSerialize();
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:135]: `alwaysSerialize()` is used, this goes in tandem with [this problem](https://github.com/elastic/elasticsearch/pull/98268/files#r1288656512). We have two different mapping semantics depending on the index version, so it's hard to tell if we have a correct configuration unless we make `index` explicit.

On the bright side, this will make it explicit to the users when obtaining mappings from indices created on previous versions vs having to understand whether the field is being indexed or not when it's not present.

**Comment by carlosdelest** [comment]
I keep having bwc error, but nothing that I can trace back to this code. Any help would be welcome 🙏

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 11]
```diff
@@ -996,11 +1015,33 @@ public static <T extends Enum<T>> Parameter<T> restrictedEnumParam(
...
+         * @param initializer     a function that reads the parameter value from an existing mapper
+         * @param defaultValue    a supplier for the default value, to be used if the parameter is undefined in a mapping
+         * @param enumClass       the enumeration class the parameter takes values from
+         * @param acceptedValues  the set of values that the parameter can take
+         */
+        public static <T extends Enum<T>> Parameter<T> restrictedEnumParam(
+            String name,
+            boolean updateable,
+            Function<FieldMapper, T> initializer,
+            Supplier<T> defaultValue,
```
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:1036]: Nitpick: Name (see above comment)

**Review by kderusso — COMMENTED** [review]
Nice work Carlos, this is looking really good!

**Comment by benwtrent** [comment]
> I keep having bwc test failures, but nothing that I can trace back to these changes. Any help would be welcome 🙏


Update your branch to latest main to make sure.

The failures may or may not be due to something in a back port branch that isn't in your branch

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 12]
```diff
@@ -974,6 +974,25 @@ public static <T extends Enum<T>> Parameter<T> enumParam(
...
+         * @param updateable    whether the parameter can be changed by a mapping update
+         * @param initializer   a function that reads the parameter value from an existing mapper
+         * @param defaultValue  a supplier for the default value, to be used if the parameter is undefined in a mapping
+         * @param enumClass     the enumeration class the parameter takes values from
+         */
+        public static <T extends Enum<T>> Parameter<T> enumParam(
+            String name,
+            boolean updateable,
+            Function<FieldMapper, T> initializer,
+            Supplier<T> defaultValue,
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:994]: That's not a name pattern I have seen in the code - I don't feel it necessary to add type information to the name. Happy to rename if anyone feels strongly about this.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 13]
```diff
@@ -127,20 +122,42 @@ public static class Builder extends FieldMapper.Builder {
...
+            this.similarity = Parameter.enumParam(
+                "similarity",
+                false,
+                m -> toType(m).similarity,
+                (Supplier<VectorSimilarity>) () -> indexedByDefault && indexed.getValue() ? VectorSimilarity.COSINE : null,
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:140]: We don't want to have `similarity` defined by default for older versions of the index (when `index` is `false` by default).  Otherwise, for older versions we'd get `index` being `false` but `similarity` being `cosine`, which would be an error.

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java (hunk 6 above)
Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:80]: There is no version information in this method 😞 . We can create new methods used in the superclass in order to create different mappings for this, but I wanted to have some reviews and consider alternatives first.

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java [hunk 14]
```diff
@@ -435,31 +452,83 @@ public void testInvalidParameters() {
...
             MapperParsingException.class,
-            () -> createDocumentMapper(fieldMapping(b -> b.field("type", "dense_vector").field("dims", 3).field("element_type", "bytes")))
+            () -> createDocumentMapper(IndexVersion.V_8_9_1, fieldMapping(b -> {
+                b.field("type", "dense_vector").field("dims", 3).field("index", true);
+            }))
+        );
+
+        assertThat(e.getMessage(), containsString("Field [index] requires field [similarity] to be configured"));
+
+        e = expectThrows(MapperParsingException.class, () -> createDocumentMapper(IndexVersion.V_8_9_1, fieldMapping(b -> {
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:470]: Sounds good, it's repeated quite a few times 👍

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java (hunk 6 above)
Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:81]: I added [a PR](https://github.com/carlosdelest/elasticsearch/pull/6) to this one with the differences for including a different method in `testDeprecatedBoost()` that would get rid of this - LMKWYT!

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 15]
```diff
@@ -127,20 +122,42 @@ public static class Builder extends FieldMapper.Builder {
...
+                VectorSimilarity.class
+            ).acceptsNull().setSerializerCheck((id, ic, v) -> v != null);
+            this.indexed.addValidator(v -> {
+                if (v) {
+                    if (similarity.getValue() == null) {
+                        throw new IllegalArgumentException("Field [index] requires field [similarity] to be configured");
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:146]: This should indicate that similarity must be configured and not null, it could be they configured it, just that they configured it to be `null`.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 16]
```diff
@@ -127,20 +122,42 @@ public static class Builder extends FieldMapper.Builder {
...
+                    }
+                } else {
+                    if (similarity.isConfigured()) {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:149]: This should be `configured` and `not null` correct? They can configure it and set it as `null` right?

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java (hunk 6 above)
Comment (benwtrent) [reply server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:81]: I referenced in that PR, but it seems to me we need a new `minimalMapping` method that accepts a `IndexVersion` parameter. It can default to ignoring that parameter and just call the existing `minimalMapping` method. 

Then tests in `MapperTestCase` that need minimal mappings referring to a version (like deprecated boosts), they can pass them.

**Review by benwtrent — COMMENTED** [review]
Very good stuff. I like many of the changes.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java (hunk 15 above)
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:146]: Good catch, thanks!

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java (hunk 16 above)
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:149]: That is correct, thanks!

File: server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java (hunk 6 above)
Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapperTests.java:81]: That's another option I checked, and it makes more sense 👍 

If we make the `index` parameter explicit though, it is not necessary. As we both think it's better to have `index` explicit, I think we can keep the code simpler this way and not introduce additional methods in `MapperTestCase`.

Happy to revisit if you think it's worth including 👍

**Comment by carlosdelest** [comment]
I'm at a loss with BWC tests. I thought it could be that `index` should now be explicit for `dense_vector` types to avoid mixed clusters getting confused about that (although it should be taken care of with the `IndexVersion`). I created this [draft PR](https://github.com/elastic/elasticsearch/pull/98379) with those changes for a quick check, but no luck.

Tests work on my machine ™️  when I try to reproduce the failing gradle command.

Any ideas about what might be going on?

**Comment by benwtrent** [comment]
@carlosdelest 🤔 I can help dig in. Have you looked at the cluster logs for the failing tests?

**Comment by carlosdelest** [comment]
> @carlosdelest 🤔 I can help dig in. Have you looked at the cluster logs for the failing tests?

I have, but not sure why it's failing. I see some inconsistent failures around:

> java.lang.RuntimeException: Failure at [aggregations/filter:104]: NodeSelector [version ranges [[8.10.0 - 8.10.0]]] rejected all nodes, living [[host=http://127.0.0.1:43963, bound=

([here](https://gradle-enterprise.elastic.co/s/jjzk4jvfnopms/tests/:qa:mixed-cluster:v8.9.1%23mixedClusterTest/org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT/test%20%7Bp0=aggregations%2Ffilter%2FStandard%20queries%20get%20cached%7D?top-execution=1))

... and some errors on
[elasticsearch-ci/bwc](https://gradle-enterprise.elastic.co/s/kvm5zx5uqtpoy)
Test failures:
[SearchWithMinCompatibleSearchNodeIT » testMinVersionAsNewVersion](https://gradle-enterprise.elastic.co/s/kvm5zx5uqtpoy/tests/:qa:mixed-cluster:v8.9.1%23mixedClusterTest/org.elasticsearch.backwards.SearchWithMinCompatibleSearchNodeIT/testMinVersionAsNewVersion)
[SearchWithMinCompatibleSearchNodeIT » testMinVersionAsOldVersion](https://gradle-enterprise.elastic.co/s/kvm5zx5uqtpoy/tests/:qa:mixed-cluster:v8.9.1%23mixedClusterTest/org.elasticsearch.backwards.SearchWithMinCompatibleSearchNodeIT/testMinVersionAsOldVersion)
[SearchWithMinCompatibleSearchNodeIT » testCcsMinimizeRoundtripsIsFalse](https://gradle-enterprise.elastic.co/s/kvm5zx5uqtpoy/tests/:qa:mixed-cluster:v8.9.1%23mixedClusterTest/org.elasticsearch.backwards.SearchWithMinCompatibleSearchNodeIT/testCcsMinimizeRoundtripsIsFalse)

The only thing that I had from the cluster logs was the following error:

> failed to commit cluster state version [1] org.elasticsearch.cluster.coordination.FailedToCommitClusterStateException: node is no longer master for term 2 while handling publication

which then turns into connect exceptions.

**Comment by benwtrent** [comment]
Digging into the logs @carlosdelest and I found the failure:

```
» [2023-08-11T09:59:06,583][WARN ][o.e.c.c.ClusterBootstrapService] [v8.9.1-1] this node is locked into cluster UUID [eOYyZC5aR96DW_VP2I09DQ] but [cluster.initial_master_nodes] is set to [v8.9.1-0, v8.9.1-1, v8.9.1-2, v8.9.1-3]; remove this setting to avoid possible data loss caused by subsequent cluster bootstrap attempts; for further information see https://www.elastic.co/guide/en/elasticsearch/reference/master/important-settings.html#initial_master_nodes
» [2023-08-11T09:59:15,724][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [v8.9.1-1] fatal error in thread [elasticsearch[v8.9.1-1][clusterApplierService#updateTask][T#1]], exiting java.lang.AssertionError: provided source [{"_doc":{"properties":{"vector1":{"type":"dense_vector","dims":768,"index":true,"similarity":"l2_norm"},"vector2":{"type":"dense_vector","dims":1024,"index":true,"similarity":"dot_product"},"vector3":{"type":"dense_vector","dims":100}}}}] differs from mapping [{"_doc":{"properties":{"vector1":{"type":"dense_vector","dims":768,"index":true,"similarity":"l2_norm"},"vector2":{"type":"dense_vector","dims":1024,"index":true,"similarity":"dot_product"},"vector3":{"type":"dense_vector","dims":100,"index":false}}}}]
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.index.mapper.DocumentMapper.<init>(DocumentMapper.java:42)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.index.mapper.MapperService.newDocumentMapper(MapperService.java:389)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.index.mapper.MapperService.updateMapping(MapperService.java:298)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.index.IndexService.updateMapping(IndexService.java:752)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.indices.cluster.IndicesClusterStateService.createIndicesAndUpdateShards(IndicesClusterStateService.java:494)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.indices.cluster.IndicesClusterStateService.applyClusterState(IndicesClusterStateService.java:251)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:539)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateAppliers(ClusterApplierService.java:525)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:498)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:429)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:154)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:916)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:217)
»       at org.elasticsearch.server@8.10.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:183)
»       at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
»       at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
»       at java.base/java.lang.Thread.run(Thread.java:1623)
```

This is because no matter the index created version, this PR is serializing the "index" default value. We should instead only do that if the index creation value is the latest one according to this change.

File: test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java [hunk 17]
```diff
@@ -86,8 +86,21 @@
...
+
     protected abstract void minimalMapping(XContentBuilder b) throws IOException;
 
+    /**
+     * Provides a way for subclasses to define minimal mappings for previous index versions
+     * @param b content builder to use for building the minimal mapping
+     * @param indexVersion index version the mapping should be created for
+     * @throws IOException
+     */
+    protected void minimalMapping(XContentBuilder b, IndexVersion indexVersion) throws IOException {
```
Comment (carlosdelest) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:100]: As discussed, I've added a new `minimalMapping()` method that takes into account index version. This is used in tests that use previous indices (`testDeprecatedBoost`) and can be overridden by subclasses if needed

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 18]
```diff
@@ -127,20 +122,46 @@ public static class Builder extends FieldMapper.Builder {
...
-            this.indexed.requiresParameter(similarity);
-            this.similarity.setSerializerCheck((id, ic, v) -> v != null);
-            this.similarity.requiresParameter(indexed);
-            this.indexOptions.requiresParameter(indexed);
-            this.indexOptions.setSerializerCheck((id, ic, v) -> v != null);
+            final boolean indexedByDefault = indexVersionCreated.onOrAfter(INDEXED_BY_DEFAULT_INDEX_VERSION);
+            this.indexed = Parameter.indexParam(m -> toType(m).indexed, indexedByDefault);
+            if (indexedByDefault) {
+                // Only serialize on newer index versions to prevent breaking existing indices when upgrading
+                this.indexed.alwaysSerialize();
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:138]: This was changed to fix bwc issues - upgrading old indices meant that this was always serialized, and thus didn't match to the stored index.

**Comment by carlosdelest** [comment]
> This is because no matter the index created version, this PR is serializing the "index" default value. We should instead only do that if the index creation value is the latest one according to this change.

Thank you so much @benwtrent ! That did the trick - and I learned a few things about looking for bwc failures.

`number_of_beers_carlos_owes_ben++`

**Comment by benwtrent** [comment]
Just to double check BWC, I am tagging to run the fully BWC suite. This will run against all 8.minor and 7.last. Hopefully catching anything else that would be weird.

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Review by benwtrent — APPROVED** [review]
:shipit:
